### ISSUE TITLE
How to debug def trainG(self, img_a, att_a, att_a_, att_b, att_b_): ???

### ISSUE THREAD

**Issue by c1a1o1** [elvisyjlin/AttGAN-PyTorch#11]
How to debug def trainG(self, img_a, att_a, att_a_, att_b, att_b_): ???

**Comment by elvisyjlin** [comment]
I cannot understand your question. What problems did you encounter?

**Comment by c1a1o1** [comment]
When I debug, the function def trainG is automatically ignored？？？

**Comment by elvisyjlin** [comment]
It is called [here](https://github.com/elvisyjlin/AttGAN-PyTorch/blob/master/train.py#L164). After several updates of D (normally it is 5 times), G is then updated.

**Comment by c1a1o1** [comment]
self.G.train()
         self.D.train()
What will these two functions do?
Thank you for your reply.

**Comment by elvisyjlin** [comment]
If you have read the [paper](https://arxiv.org/pdf/1711.10678.pdf), the overall objective functions are listed in Section III, Overall Objective. on the fifth page.
* [trainD()](https://github.com/elvisyjlin/AttGAN-PyTorch/blob/master/attgan.py#L194) takes a pair of real images A and B. It generates a fake image A' from A with B's attributes. Then we calculate the adversarial loss with gradient penalty and the classification loss to update D's weigths. See formula 15.
* [trainG()](https://github.com/elvisyjlin/AttGAN-PyTorch/blob/master/attgan.py#L165) also takes a pair of A and B and generates A'. We compute the adversarial loss, the classification loss, and the reconstruction loss to update G's weights. See formula 14.

A GAN evolves like a min-max game, where D tries to maximize an objective function whilst G attempts to minimize it. So they are updated in turn and help each other grow iteratively. If you are not familiar with GAN, please read some articles or papers about it.

However, if you just use a pretrained model, these functions are not called at all. Because the model has been trained to generate plausible human face images. There is no need to update the weights of the model.

**Comment by c1a1o1** [comment]
Thank you for your reply. I get it！

**Comment by elvisyjlin** [comment]
You’re welcome! By the way, have you found out what the problem is in the other issue you opened?

**Comment by c1a1o1** [comment]
Thank you for your reply. I fixed it！

**Comment by elvisyjlin** [comment]
I’m glad to hear that!
